integrations/db-mongodb/src/MongoDb.ts

```diff
@@ -0,0 +1,103 @@
+  /** The name of the database we want to use. If not provided, use database name from connection string. A new database is created if doesn't exist yet. */
+  databaseName?: string;
+  /** A new collection is created with that name if doesn't exist yet. */
+  collectionName?: string;
+}
+
+export type MongoDbInitConfig = RequiredOnlyWhere<MongoDbConfig, 'connectionString'>;
+
+export interface MongoDbItem {
+  id: string;
+  user?: PersistableUserData;
+  session?: PersistableSessionData;
+  createdAt?: string;
+  updatedAt?: string;
+}
+
+/** Default database name in MongoDB. if no name was specified, 'test' is used. See https://docs.mongodb.com/manual/tutorial/getting-started/#getting-started */
+export const MONGODB_DEFAULT_DATABASE_NAME = 'test';
+export const JOVO_DEFAULT_DATABASE_NAME = 'jovo_db';
+export const JOVO_DEFAULT_COLLECTION_NAME = 'jovoUsers';
+
+export class MongoDb extends DbPlugin<MongoDbConfig> {
+  /** A single client promise to be shared by Jovo and others components following MongoDB best practice: https://docs.atlas.mongodb.com/best-practices-connecting-from-aws-lambda/#connection-examples */
+  readonly client: Promise<MongoClient> = new MongoClient(this.config.connectionString).connect();
+
+  constructor(config: MongoDbInitConfig) {
+    super(config);
+  }
+
+  getDefaultConfig(): MongoDbConfig {
+    return {
+      ...super.getDefaultConfig(),
+      connectionString: '<YOUR-MONGODB-URI>',
+      databaseName: JOVO_DEFAULT_DATABASE_NAME,
+      collectionName: JOVO_DEFAULT_COLLECTION_NAME,
+    };
+  }
+
+  getInitConfig(): MongoDbInitConfig {
+    return { connectionString: '<YOUR-MONGODB-URI>' };
+  }
+
+  mount(parent: HandleRequest): Promise<void> | void {
+    super.mount(parent);
+  }
+
+  async initialize(): Promise<void> {
+    if ((await this.jovoDb()).databaseName === MONGODB_DEFAULT_DATABASE_NAME) {
+      // eslint-disable-next-line no-console
+      console.warn('[MongoDB] Warning: The "test" database is being used.');
+    }
+  }
+
+  async loadData(userId: string, jovo: Jovo): Promise<void> {
+    const users = await this.jovoUsers();
+    const filter = { _id: userId };
+    const dbItem = (await users.findOne(filter)) as DbItem;
+    if (dbItem) {
+      jovo.$user.isNew = false;
+      jovo.setPersistableData(dbItem, this.config.storedElements);
+    }
+  }
+
+  async saveData(userId: string, jovo: Jovo): Promise<void> {
+    const users = await this.jovoUsers();
+    const item: DbItem = { _id: userId };
+    await this.applyPersistableData(jovo, item);
+    const filter = { _id: userId };
+    await users.updateOne(filter, { $set: item }, { upsert: true });
+  }
+
+  /** MongoDB creates the database if doesn't exist yet */
+  async jovoDb(): Promise<Db> {
+    const connection = await this.client;
+    if (this.config.databaseName) {
+      return connection!.db(this.config.databaseName);
+    } else {
+      //If not provided, use database name from connection string.
```

Comment:
You are right. It's not entering there.

That's an ordinary check I put there before adding default database name in getDefaultConfig.

Since the database name can be specified either in the connection URL or in the MongoDb integration constructor (`databaseName`), I'm not sure what to do here.

If a user specify the DB name in the connection url and default config for DB name `databaseName` is present, it has no efect and will never connect to DB specified in connection URL. That's a problem I guess.

With the new client for MongoDb for Node, I can't find an elegant way to check either if DB name was specified in URL or not.

I would remove `databaseName` from `getDefaultConfig()` and set it as optional in the constructor, and leave the default collectionName like it is now. This way, a user can set the DB name either in the URL or in the constructor. By doing that, the check you referred would make sense.

Another option is to not allow a user to specify a dataBase name in the URL, but in the `databaseName` argument in the constructor. That would have to be clear in the docs to avoid future issues.

Reply:
Now the behaviour is:

- `databaseName` argument in MongoDb constructor takes precedence over database name in connection URL.
- If no `databaseName` is specified, it will connect to database specified in URL.
- If no database is specified in the URL nor in the databaseName, `test` database is used (deault MongoDb behaviour).

So:

1.
```
    new MongoDb({
      connectionString: 'mongodb+srv://userX@clusterX.mongodb.net/mydb',
      databaseName: 'mydb2'
    }),
``` 
`mydb2` will be used.


2.
```
    new MongoDb({
      connectionString: 'mongodb+srv://userX@clusterX.mongodb.net/mydb',
    }),
``` 
`mydb` will be used.


3.
```
    new MongoDb({
      connectionString: 'mongodb+srv://userX@clusterX.mongodb.net/',
    }),
``` 
Shows a warning in app instantiation:
`[MongoDB] Warning: The "test" database is being used.`



Please tell me what you think.

Reply:
I like the new solution! Thank you 🎉